From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp60: Implement locale::operator() for strings.

msvcp60's locale::operator()(string, string): whether the first string
collates before the second by the locale's collate<char> facet.
---
 dlls/msvcp60/msvcp60.spec |  6 +++---
 dlls/msvcp90/locale.c     | 19 +++++++++++++++++++
 2 files changed, 22 insertions(+), 3 deletions(-)

diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index 0bc26b1..e18498c 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -2154,9 +2154,9 @@
 @ cdecl -arch=win64 ??Pstd@@YA_NPEBDAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@0@@Z(ptr ptr) MSVCP_basic_string_char_geq_cstr_bstr
 @ cdecl -arch=win32 ??Pstd@@YA_NPBGABV?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@0@@Z(ptr ptr) MSVCP_basic_string_wchar_geq_cstr_bstr
 @ cdecl -arch=win64 ??Pstd@@YA_NPEBGAEBV?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@0@@Z(ptr ptr) MSVCP_basic_string_wchar_geq_cstr_bstr
-@ stub -arch=arm ??Rlocale@std@@QBA_NABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@0@Z
-@ stub -arch=i386 ??Rlocale@std@@QBE_NABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@0@Z
-@ stub -arch=win64 ??Rlocale@std@@QEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@0@Z
+@ cdecl -arch=arm ??Rlocale@std@@QBA_NABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@0@Z(ptr ptr ptr) locale_operator_call
+@ thiscall -arch=i386 ??Rlocale@std@@QBE_NABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@0@Z(ptr ptr ptr) locale_operator_call
+@ cdecl -arch=win64 ??Rlocale@std@@QEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@0@Z(ptr ptr ptr) locale_operator_call
 @ cdecl -arch=arm ??X?$_Complex_base@M@std@@QAAAAV01@ABM@Z(ptr ptr) complex_float_mult_assign_float
 @ thiscall -arch=i386 ??X?$_Complex_base@M@std@@QAEAAV01@ABM@Z(ptr ptr) complex_float_mult_assign_float
 @ cdecl -arch=win64 ??X?$_Complex_base@M@std@@QEAAAEAV01@AEBM@Z(ptr ptr) complex_float_mult_assign_float
diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
index 3920ff2..ad841e3 100644
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -16918,3 +16918,22 @@ locale_id* __thiscall locale_id_op_assign(locale_id *this, const locale_id *rhs)
     return this;
 }
 #endif
+
+#if _MSVCP_VER < 70
+/* ??Rlocale@std@@QBE_NABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@0@Z */
+/* ??Rlocale@std@@QEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@0@Z */
+/* whether s1 comes before s2 by the locale's collate<char> facet */
+DEFINE_THISCALL_WRAPPER(locale_operator_call, 12)
+bool __thiscall locale_operator_call(const locale *this, const basic_string_char *s1, const basic_string_char *s2)
+{
+    const collate *coll = collate_char_use_facet(this);
+    const char *p1 = MSVCP_basic_string_char_c_str(s1), *p2 = MSVCP_basic_string_char_c_str(s2);
+    size_t len1 = MSVCP_basic_string_char_length(s1), len2 = MSVCP_basic_string_char_length(s2);
+
+    TRACE("(%p %s %s)\n", this, debugstr_a(p1), debugstr_a(p2));
+#ifdef SG_MUTANT_LOCALE_CALL_GREATER
+    return collate_char_compare(coll, p1, p1 + len1, p2, p2 + len2) > 0;
+#endif
+    return collate_char_compare(coll, p1, p1 + len1, p2, p2 + len2) < 0;
+}
+#endif
